plugins/application/elasticsearch/main.go

```diff
@@ -126,21 +132,66 @@ func (es *Elasticsearch) Run(ctx context.Context, done chan bool) {
 		es.logger.Info("removed indices")
 	}
 
-	for {
-		select {
-		case <-ctx.Done():
-			goto done
-		case dumped := <-es.dump:
-			if err := es.client.Index(dumped.index, dumped.record, es.configuration.BulkIndex); err != nil {
-				es.logger.Metadata(logging.Metadata{"plugin": appname, "event": dumped.record, "error": err})
-				es.logger.Error("failed to index event - disregarding")
-			} else {
-				es.logger.Debug("successfully indexed document(s)")
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	// spawn buffer flusher
+	go func(es *Elasticsearch, ctx context.Context, wg *sync.WaitGroup) {
+		defer wg.Done()
+		tick := time.NewTicker(time.Duration(es.configuration.BufferTimeout) * time.Second)
+
+		for {
+			select {
+			case <-ctx.Done():
+				tick.Stop()
+				es.logger.Metadata(logging.Metadata{"plugin": appname})
+				es.logger.Debug("shutting down buffer flusher")
+				return
+			case <-tick.C:
+				es.bufferMutex.Lock()
+				for index, record := range es.buffer {
+					if err := es.client.Index(index, record, es.configuration.BulkIndex); err != nil {
+						es.logger.Metadata(logging.Metadata{"plugin": appname, "records": len(record), "index": index, "error": err})
+						es.logger.Error("failed to flush buffer - disregarding")
+					} else {
+						es.logger.Metadata(logging.Metadata{"plugin": appname, "records": len(record), "index": index})
+						es.logger.Debug("successfully flushed buffer")
+					}
+					delete(es.buffer, index)
+				}
+				es.bufferMutex.Unlock()
 			}
 		}
+	}(es, ctx, &wg)
+
+	// spawn index workers
+	for i := 0; i < es.configuration.IndexWorkers; i++ {
```

Comment:
I'm confused by the history here. On July 13th Paul suggested eliminating the worker pool by re-using the goroutine already spawned by the call ReceiveEvent(). On July 14th you said "That was massive performance improvement indeed Paul!"  but I still see the workers code here. Did this approach need to be adandoned?